### PR TITLE
Add Advanced Expression Simplification System

### PR DESCRIPTION
**Pull request by jeffreyhorn**
This PR adds a comprehensive expression simplification system with configurable modes and advanced algebraic simplifications for automatic differentiation expressions.

### Overview

Implements a three-tier simplification system (`none`, `basic`, `advanced`) with `advanced` as the default mode. Advanced simplification includes additive term collection, multiplicative term collection, multiplicative cancellation, and power rules.

#### 1. **Configuration-Based Simplification Modes**
- Added `simplification` config option with three modes:
  - `none`: No simplification
  - `basic`: Apply only basic algebraic rules (existing behavior)
  - `advanced`: Apply basic rules plus term collection and advanced patterns (new default)
- Integrated into `gradient.py` and `constraint_jacobian.py` pipelines
- Full validation with helpful error messages

#### 2. **Additive Term Collection**
Simplifies addition and subtraction patterns:
- Constant collection: `1 + x + 1 → x + 2`
- Like-term collection: `x + y + x + y → 2*x + 2*y`
- Coefficient collection: `2*x + 3*x → 5*x`
- Term cancellation: `x - x → 0`, `x + y - x → y`
- Complex bases: `x*y + 2*x*y → 3*x*y`

#### 3. **Multiplicative Term Collection**
Simplifies multiplication patterns:
- Variable collection: `x * x → x^2`, `x * x * x → x^3`
- Power multiplication: `x^2 * x^3 → x^5`
- Mixed multiplication: `x^2 * x → x^3`, `x * x^2 → x^3`

#### 4. **Multiplicative Cancellation**
Simplifies division patterns:
- Basic cancellation: `2*x / 2 → x`
- Evaluated cancellation: `2*x / (1+1) → x`
- Works with left and right factors: `(c*expr)/c → expr`, `(expr*c)/c → expr`

#### 5. **Power Simplification Rules**
Applies algebraic identities for exponents:
- Power multiplication: `x^a * x^b → x^(a+b)`
- Power division: `x^5 / x^2 → x^3`, `x / x^2 → 1/x`
- Nested powers: `(x^2)^3 → x^6`
- Edge cases: Handles cancellation to 1, negative exponents

### Files Changed

**Core Implementation:**
- `src/config.py` - Added `simplification` field with validation
- `src/ad/ad_core.py` - Added `apply_simplification()` helper and enhanced `simplify_advanced()` with special cases for `*`, `/`, `**` operations
- `src/ad/gradient.py` - Updated to use config-based simplification
- `src/ad/constraint_jacobian.py` - Updated to use config-based simplification
- `src/ad/term_collection.py` - Added `simplify_multiplicative_cancellation()` and `simplify_power_rules()` functions

**Tests (75 new tests, all passing):**
- `tests/unit/test_config.py` - 7 tests for config validation
- `tests/unit/ad/test_apply_simplification.py` - 8 tests for mode-specific behavior
- `tests/unit/ad/test_multiplicative_cancellation.py` - 19 tests for cancellation patterns
- `tests/unit/ad/test_power_simplification.py` - 31 tests for power rules
- Total AD tests: **375 (up from 300)**

**Documentation:**
- `README.md` - Added Expression Simplification section with examples
- `docs/USER_GUIDE.md` - Comprehensive guide with 11 numbered simplification patterns
- `CHANGELOG.md` - Documented all new features

### Testing

All 375 automated differentiation tests pass, including:
- Existing tests verify no regression
- 75 new tests cover all new simplification patterns
- Edge cases: zero exponents, negative exponents, zero denominators, identity operations

### Example Usage

```python
from src.config import Config
from src.ad.gradient import compute_gradient

# Default: advanced simplification
gradient = compute_gradient(expr, variables)

# Disable simplification
config = Config(simplification="none")
gradient = compute_gradient(expr, variables, config)

# Basic simplification only
config = Config(simplification="basic")
gradient = compute_gradient(expr, variables, config)
```

### Before/After Examples

```
# Additive term collection
1 + x + 1                    → x + 2
x + y + x + y                → 2*x + 2*y

# Multiplicative term collection
x * x * x                    → x^3
x^2 * x^3                    → x^5

# Cancellation
2*x / (1+1)                  → x

# Power rules
x^5 / x^2                    → x^3
(x^2)^3                      → x^6
```

---

This PR significantly improves the readability and compactness of generated gradient and Jacobian expressions while maintaining mathematical correctness.

### CODE REVIEW

File: src/ad/ad_core.py [hunk 1]
```diff
@@ -266,3 +266,160 @@ def simplify(expr: Expr) -> Expr:
...
+        >>> expr = Binary("+", Binary("+", Const(1), VarRef("x", ())), Const(1))
+        >>> result = simplify_advanced(expr)
+        >>> # Result: Binary("+", VarRef("x", ()), Const(2)) or Const(2) + VarRef("x")
+    """
+    from ..ir.ast import Binary, Call, Sum, Unary
+    from .term_collection import (
+        collect_like_terms,
+        simplify_multiplicative_cancellation,
+        simplify_power_rules,
+    )
```
[the remark this answers is not in the document]
Comment (jeffreyhorn) [reply src/ad/ad_core.py]: - Moved imports from inside `simplify_advanced()` to module-level
- Avoids repeated import overhead during differentiation of large expressions
- Cleaner code structure
